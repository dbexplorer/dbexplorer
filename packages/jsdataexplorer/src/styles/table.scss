.jsde-table {
    height: 0;
}

.jsde-table-head {
    color: white;
    background-color: black;
    position: sticky;
    top: 0;
}

.jsde-table-row {
    &:hover {
        background-color: gray;
    }
}

.jsde-table-row--explored {
    background-color: lightgrey;
}

.jsde-table-cell {
    padding: 8px;
    white-space: nowrap;
}