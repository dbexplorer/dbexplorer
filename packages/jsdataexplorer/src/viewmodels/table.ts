import { IGetDataOptions } from "../data";
import { cssPrefix, keysEqual } from "../utils";

export class TableCellViewModel {
  private text: string;
  getText() {
    return this.text;
  }
  setText(text: string) {
    this.text = text;
    this.updateCallback(this.text);
  }
  constructor(text: string = null) {
    this.text = text;
  }
  css() {
    return { root: cssPrefix("table-cell") };
  }
  updateCallback: (data: string) => any = () => { };
}
export class TableRowViewModel {
  private cells: TableCellViewModel[];
  private explored: boolean;
  getKey() {
    return this.key as string;
  }
  getCells() {
    return this.cells;
  }
  setCellsText(data: string[]) {
    if (!this.cells) {
      this.cells = data.map((s) => new TableCellViewModel(s));
    }
    else {
      data.map((s, idx) => this.cells[idx].setText(s));
    }
  }
  setExplored(explored: boolean, needUpdate: boolean = true) {
    this.explored = explored;
    if (needUpdate) this.updateCssCallback(this.css().root);
  }

  cssModifiers() {
    return {
      root: {
        explored: "--explored"
      }
    };
  }
  css() {
    return { root: cssPrefix("table-row") + (this.explored ? " " + cssPrefix("table-row") + this.cssModifiers().root.explored : "") };
  }
  constructor(data: string[], private key: string | string[]) {
    this.setCellsText(data);
  }
  updateDataCallback: (data: TableCellViewModel[]) => any;
  updateCssCallback: (css: string) => any;
  exploreCallback: () => void;
}

export class TableHeaderViewModel {
  public captionsViewModel: TableRowViewModel;
  constructor(captions: string[]) {
    this.captionsViewModel = new TableRowViewModel(captions, null);
  }
}

export class TableColumnDescription {
  name: string;
  title?: string;
  hasReference?: boolean;
  isMain?: boolean;
}

export class TableViewModel {
  private columns: TableColumnDescription[];
  private dataOffset: number = 0;
  private dataBackOffset: number = 0;
  private exploredRow: TableRowViewModel = null;
  private loadBackVisible = true;
  dataPartRowCount: number = 10;
  headerViewModel: TableHeaderViewModel;
  private setLoadBackVisible(value: boolean) {
    this.loadBackVisible = value;
    this.loadBackVisibleCallback(false);
  }
  loadData(back = false) {
    const options = { limit: this.dataPartRowCount, offset: back ? this.dataBackOffset : this.dataOffset, key: this.key, back: back };

    if (back && !this.key) {
      this.setLoadBackVisible(false);
      return;
    }

    this.getDataCallback(
      options,
      (data: any) => {
        if (back) {
          this.dataBackOffset += this.dataPartRowCount;
        }
        else {
          this.dataOffset += this.dataPartRowCount;
        }
        if (data.length > 0) {
          const rows = data.map(
            (rowData: any) => {
              const row = new TableRowViewModel(
                this.columns.map((col) => rowData.data[col.name].toString()),
                rowData.key
              )
              if (keysEqual(rowData.key, this.key)) {
                this.exploredRow = row;
                this.exploredRow.setExplored(true, false);
                this.exploreRowCallback(row);
              }
              row.exploreCallback = () => {
                this.exploredRow?.setExplored(false);
                const newExploredRow = (this.exploredRow !== row) ? row : null;
                this.exploredRow = newExploredRow;
                this.exploredRow?.setExplored(true);
                this.exploreRowCallback(newExploredRow);
              }
              return row;
            }
          )
          if (back) {
            this.rows = rows.reverse().concat(this.rows);
          }
          else {
            this.rows = this.rows.concat(rows);
          }
          this.addRowsCallback(this.rows);
        }
        else {
          if (back) this.setLoadBackVisible(false); else this.loadMoreVisibleCallback(false);
        }
        if (!this.key) this.loadBackVisibleCallback(false);
      }
    );
  }
  rows: TableRowViewModel[];
  css() {
    return {
      root: cssPrefix("table"),
      head: cssPrefix("table-head"),
      body: cssPrefix("table-body"),
      foot: cssPrefix("table-foot"),
      loadingTop: cssPrefix("table-loading-top"),
      loadingBottom: cssPrefix("table-loading-bottom"),
    };
  }
  getDataCallback: (options: IGetDataOptions, ready: any) => void;
  constructor(columns: TableColumnDescription[], private key: any = undefined, private title = "Table") {
    this.columns = columns;
    this.headerViewModel = new TableHeaderViewModel(
      columns.map((col) => col.title || col.name)
    );
    this.rows = [];
  }
  getKey() {
    return this.key;
  }
  getKeyAsString() {
    return this.key.toString();
  }
  getTitle() {
    return this.title;
  }
  getCurrentRowIndex() {
    let index = this.rows.indexOf(this.rows.filter(r => keysEqual(r.getKey(), this.key))[0]);
    if (index >= 0 && this.loadBackVisible) index++;
    return index;
  }
  getLoadBackVisible(): any {
    return true;
  }
  getLoadMoreVisible(): any {
    return true;
  }
  addRowsCallback: (rows: TableRowViewModel[]) => any;
  loadMoreVisibleCallback: (visible: boolean) => any;
  loadBackVisibleCallback: (visible: boolean) => any = () => { };
  exploreRowCallback: (row: TableRowViewModel) => any;
}
