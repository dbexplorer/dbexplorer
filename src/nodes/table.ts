export class TableCell {
  constructor() {}
}

export class TableRow {
  constructor() {}
}

export class TableNode {
  constructor() {}
}
