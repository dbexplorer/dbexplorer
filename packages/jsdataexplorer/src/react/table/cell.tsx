import React, { useState, useEffect } from 'react';
import { TableCellViewModel } from '../../viewmodels/table';

export function TableCell({ model }: { model: TableCellViewModel }) {
  const [text, setText] = useState(model.getText());

  model.updateCallback = setText;

  return <td className={model.css().root}>{text}</td>;
}