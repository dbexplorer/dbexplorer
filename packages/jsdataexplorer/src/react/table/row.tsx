import React, { useState } from 'react';
import { TableRowViewModel } from '../../viewmodels/table';
import { TableCell } from './cell';

export function TableRow({ model }: { model: TableRowViewModel }) {
  const [rootCss, setRootCss] = useState(model.css().root);

  model.updateCssCallback = setRootCss;
  function handleClick() {
    return model.exploreCallback();
  }
  return (
    <tr className={rootCss} onClick={handleClick}>{model.getCells().map((cell, index) => <TableCell key={index} model={cell} />)}</tr>
  );
}