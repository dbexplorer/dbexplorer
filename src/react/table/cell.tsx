import React from 'react';
import { TableCellViewModel } from '../../viewmodels/table';

interface IProps {
  model: TableCellViewModel;
}

interface IState {
  text: string;
}
export class TableCell extends React.Component<IProps, IState> {
  constructor(props: any) {
    super(props);
    this.state = { text: props.model.getText() };
    props.model.updateCallback = (data: string) => {
      this.setState({ text: data });
    };
  }
  render() {
    return (
      <td>{this.state.text}</td>
    );
  }
}