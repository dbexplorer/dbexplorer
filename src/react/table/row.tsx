import React from 'react';
import { TableRowViewModel } from '../../viewmodels/table';
import { TableCell } from './cell';

interface IProps {
  model: TableRowViewModel;
}

interface IState {
}
export class TableRow extends React.Component<IProps, IState> {
  constructor(props: any) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }
  getCells() {
    return this.props.model.getCells();
  }
  handleClick() {
    this.props.model.exploreCallback();
  }
  render() {
    return (
      <><tr onClick={this.handleClick}>{this.getCells().map((cell, index) => <TableCell key={index} model={cell} />)}</tr></>
    );
  }
}