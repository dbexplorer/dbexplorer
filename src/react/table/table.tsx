import React from 'react';
import { TableRowViewModel, TableViewModel } from '../../viewmodels/table';
import { TableRow } from './row';

interface IProps {
  model: TableViewModel;
}

interface IState {
  rows: TableRowViewModel[];
}
export class Table extends React.Component<IProps, IState> {
  constructor(props: any) {
    super(props);
    this.state = { rows: props.model.rows };
    this.props.model.addRowsCallback = (data: []) => {
      this.setState({ rows: data });
    };
    this.handleLoadMoreClick = this.handleLoadMoreClick.bind(this);
  }
  getHeaderCells() {
    return this.props.model.headerViewModel.captionsViewModel.getCells();
  }
  handleLoadMoreClick() {
    this.props.model.loadData();
  }

  componentDidMount() {
    this.props.model.loadData();
  }
  render() {
    return (
      <table>
        <thead><tr>{this.getHeaderCells().map((cell, index) => <th key={index}>{cell.getText()}</th>)}</tr></thead>
        <tbody>
          {
            this.state.rows.map((row, index) => <TableRow key={index} model={row} />)
          }
        </tbody>
        <tfoot><tr><td><button onClick={this.handleLoadMoreClick}>Load more data...</button></td></tr></tfoot>
      </table>
    );
  }
}