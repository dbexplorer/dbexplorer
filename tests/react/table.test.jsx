import React from 'react';
import { TableCell } from "../../src/react/table/cell";
import { TableRow } from "../../src/react/table/row";
import { Table } from "../../src/react/table/table";

import { render, fireEvent, waitFor, screen } from '@testing-library/react'
import {
  TableCellViewModel, TableRowViewModel, TableViewModel
} from "../../src/viewmodels/table";
import { act } from 'react-dom/test-utils';
test("Table cell test", () => {
  var tableCellViewModel = new TableCellViewModel("t1");

  const { container } = render(<TableCell model={tableCellViewModel} />);
  expect(container.firstChild).toMatchSnapshot();
  act(() => {
    tableCellViewModel.setText("t2");
  })

  expect(container.firstChild).toMatchSnapshot();
});
test("Table row test", () => {
  var tableRowViewModel = new TableRowViewModel(["t1", "t2"], "t1");

  const { container } = render(<TableRow model={tableRowViewModel} />);
  expect(container.firstChild).toMatchSnapshot();
  act(() => {
    tableRowViewModel.getCells()[0].setText("t3");
  })
  expect(container.firstChild).toMatchSnapshot();
});
test("Table test", () => {
  var tableViewModel = new TableViewModel([
    {
      name: "f1"
    },
    {
      name: "f2"
    },
    {
      name: "f3"
    }
  ]);
  tableViewModel.dataPartRowCount = 2;
  tableViewModel.getDataCallback = (limit, offset, ready) => {
    if (limit === 2) {
      ready([
        { key: 1 + offset * 2, data: { f1: 1 + offset * 2, f2: "one", f3: "first" } },
        { key: 2 + offset * 2, data: { f1: 2 + offset * 2, f2: "two", f3: "second" } }
      ]);
    }
  };

  const { container } = render(<Table model={tableViewModel} />);
  expect(container.querySelector("tbody").childElementCount).toEqual(2);
  expect(container.firstChild).toMatchSnapshot();
});
test("table test - load more", () => {
  var tableViewModel = new TableViewModel([
    {
      name: "f1"
    },
    {
      name: "f2"
    },
    {
      name: "f3"
    }
  ]);
  tableViewModel.dataPartRowCount = 2;
  tableViewModel.getDataCallback = (limit, offset, ready) => {
    if (limit === 2) {
      ready([
        { key: 1 + offset * 2, data: { f1: 1 + offset * 2, f2: "one", f3: "first" } },
        { key: 2 + offset * 2, data: { f1: 2 + offset * 2, f2: "two", f3: "second" } }
      ]);
    }
  };
  const { container } = render(<Table model={tableViewModel} />);
  act(() => {
    tableViewModel.loadData();
  })
  fireEvent.click(container.querySelector("tfoot button"));
  expect(container.firstChild).toMatchSnapshot();
});
test("Table test - row click", () => {
  var tableViewModel = new TableViewModel([
    {
      name: "f1"
    },
    {
      name: "f2"
    },
    {
      name: "f3"
    }
  ]);
  tableViewModel.dataPartRowCount = 2;
  tableViewModel.getDataCallback = (limit, offset, ready) => {
    if (limit === 2) {
      ready([
        { key: 1 + offset * 2, data: { f1: 1 + offset * 2, f2: "one", f3: "first" } },
        { key: 2 + offset * 2, data: { f1: 2 + offset * 2, f2: "two", f3: "second" } }
      ]);
    }
  };
  const { container } = render(<Table model={tableViewModel} />);
  var d = [];
  tableViewModel.exploreRowCallback = (data) => {
    d = data.getCells().map(c => { return { text: c.text } });
  }

  fireEvent.click(container.querySelector("tbody tr"));
  expect(d).toEqual([{ "text": "1" }, { "text": "one" }, { "text": "first" }]);
});
