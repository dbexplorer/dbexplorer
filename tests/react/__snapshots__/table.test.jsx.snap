// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table cell test 1`] = `
<td>
  t1
</td>
`;

exports[`Table cell test 2`] = `
<td>
  t2
</td>
`;

exports[`Table row test 1`] = `
<tr>
  <td>
    t1
  </td>
  <td>
    t2
  </td>
</tr>
`;

exports[`Table row test 2`] = `
<tr>
  <td>
    t3
  </td>
  <td>
    t2
  </td>
</tr>
`;

exports[`Table test 1`] = `
<table>
  <thead>
    <tr>
      <th>
        f1
      </th>
      <th>
        f2
      </th>
      <th>
        f3
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        1
      </td>
      <td>
        one
      </td>
      <td>
        first
      </td>
    </tr>
    <tr>
      <td>
        2
      </td>
      <td>
        two
      </td>
      <td>
        second
      </td>
    </tr>
  </tbody>
  <tfoot>
    <tr>
      <td>
        <button>
          Load more data...
        </button>
      </td>
    </tr>
  </tfoot>
</table>
`;

exports[`table test - load more 1`] = `
<table>
  <thead>
    <tr>
      <th>
        f1
      </th>
      <th>
        f2
      </th>
      <th>
        f3
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        1
      </td>
      <td>
        one
      </td>
      <td>
        first
      </td>
    </tr>
    <tr>
      <td>
        2
      </td>
      <td>
        two
      </td>
      <td>
        second
      </td>
    </tr>
  </tbody>
  <tfoot>
    <tr>
      <td>
        <button>
          Load more data...
        </button>
      </td>
    </tr>
  </tfoot>
</table>
`;
